From 00ac4e65e83282ffb4cc5f553103c6d80b2d48c5 Mon Sep 17 00:00:00 2001
From: jakub <jakub@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Wed, 2 Jul 2014 09:48:56 +0000
Subject: [PATCH 083/134] 	* decl.c (variable_decl): Reject old style
 initialization 	for derived type components.

	* gfortran.dg/oldstyle_5.f: New test.


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_8-branch@212229 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/fortran/ChangeLog                  | 6 ++++++
 gcc/fortran/decl.c                     | 7 +++++++
 gcc/testsuite/ChangeLog                | 5 +++++
 gcc/testsuite/gfortran.dg/oldstyle_5.f | 8 ++++++++
 4 files changed, 26 insertions(+)
 create mode 100644 gcc/testsuite/gfortran.dg/oldstyle_5.f

diff --git a/gcc/fortran/ChangeLog b/gcc/fortran/ChangeLog
index 8033c7a..65cfa6f 100644
--- a/gcc/fortran/ChangeLog
+++ b/gcc/fortran/ChangeLog
@@ -1,3 +1,9 @@
+2014-07-02  Jakub Jelinek  <jakub@redhat.com>
+	    Fritz Reese  <Reese-Fritz@zai.com>
+
+	* decl.c (variable_decl): Reject old style initialization
+	for derived type components.
+
 2014-06-15  Francois-Xavier Coudert  <fxcoudert@gcc.gnu.org>
 
 	Backport from trunk.
diff --git a/gcc/fortran/decl.c b/gcc/fortran/decl.c
index 7dec803..9292418a 100644
--- a/gcc/fortran/decl.c
+++ b/gcc/fortran/decl.c
@@ -1996,6 +1996,13 @@ variable_decl (int elem)
       if (gfc_notify_std (GFC_STD_GNU, "Old-style "
 			  "initialization at %C") == FAILURE)
 	return MATCH_ERROR;
+      else if (gfc_current_state () == COMP_DERIVED)
+	{
+	  gfc_error ("Invalid old style initialization for derived type "
+		     "component at %C");
+	  m = MATCH_ERROR;
+	  goto cleanup;
+	}
 
       return match_old_style_init (name);
     }
diff --git a/gcc/testsuite/ChangeLog b/gcc/testsuite/ChangeLog
index 8ccfcfe..facd6dd 100644
--- a/gcc/testsuite/ChangeLog
+++ b/gcc/testsuite/ChangeLog
@@ -1,3 +1,8 @@
+2014-07-02  Jakub Jelinek  <jakub@redhat.com>
+	    Fritz Reese  <Reese-Fritz@zai.com>
+
+	* gfortran.dg/oldstyle_5.f: New test.
+
 2014-06-30  Thomas Preud'homme  <thomas.preudhomme@arm.com>
 
 	Backport from mainline
diff --git a/gcc/testsuite/gfortran.dg/oldstyle_5.f b/gcc/testsuite/gfortran.dg/oldstyle_5.f
new file mode 100644
index 0000000..8a0d311
--- /dev/null
+++ b/gcc/testsuite/gfortran.dg/oldstyle_5.f
@@ -0,0 +1,8 @@
+C { dg-do compile }
+      TYPE T
+      INTEGER A(2)/1,2/ ! { dg-error "Invalid old style initialization for derived type component" }
+      END TYPE
+      TYPE S
+      INTEGER B/1/ ! { dg-error "Invalid old style initialization for derived type component" }
+      END TYPE
+      END
-- 
1.8.5.5

